File: C++/practice_problems.cpp
#include <iostream>
#include <cstring>
#include <cstdlib>

using namespace std;


class I_sorted{
public:
    bool iterative_sorted(int arr[], int len){
        for(int i=0; i<len - 1; i++){
            if(arr[i] < arr[i+1]){
                return false;
            }
        } return true;
    }
};

class R_occurance{
public:
    int r_occurance_number(int arr[], int len, int target){
        if(len==0)
            return 0;
        else if(len > 0){
            if(target == arr[len-1]){
                return 1 + r_occurance_number(arr, len-1, target);
            }return r_occurance_number(arr, len-1, target);
        }else
            return 0;
    }
};

class Smallest{
public:
    int i_smallest(int arr[], int len){
        int temp=arr[0];
        if(len==0)
            return 0;
        for(int i=0; i<len; i++){
            if(arr[i]<temp)
                temp = arr[i];
        } return temp;
    }

    int r_smallest(int arr[], int len){
        if(len==0)
            return 0;
        if(len==1)
            return arr[len-1];
        int min = r_smallest(arr, len-1);

        if(min<arr[len-1])
            return min;
        else
            return arr[len-1];
    }
};

class Palindrome{
public:
    bool i_palindrome(int arr[], int len){
        for(int i=0; i<len/2; i++){
            if(arr[i]!=arr[len-i-1])
                return false;
        } return true;
    }

    bool r_palindrome(int arr[], int left, int right){
        if(left>=right)
            return true;
        if(arr[left]!=arr[right])
            return false;
        else
            return r_palindrome(arr, left+1, right-1);
    }
};

class Practice_Problems{
private:
    I_sorted sortcheck;
    R_occurance occurance;
    Smallest small;
    Palindrome palindromechecker;
public:
    int menu(){
        int opt;
        cout<<"----------------------------Practice Problems-----------------------------\n";
        cout<<"1. iterative method to determine if array is sorted in descending order.\n";
        cout<<"2. recursive method to count the occurrence of a number in an array\n";
        cout<<"3. iterative method to find the smallest element in an array\n";
        cout<<"4. recursive method to find the smallest element in an array\n";
        cout<<"5. iterative method to check if the array is PALINDROME\n";
        cout<<"6. recursive method to check if the array is PALINDROME\n";
        cout<<"7. change the array\n";
        cout<<"8. Exit\n";
        cout<<"--------------------------------------------------------------------------\n";
        cout<<"[1-8]-> ";
        cin>>opt;
        return opt;
    }

    void cases(){
        const int MAX = 10;
        int arr[MAX];
        int len;
        int target;
        cout<<"Enter size of your array (max 10): ";
        cin>>len;
        cout<<"Enter elements: ";
        for(int i=0; i<len; i++) {
            cin>>arr[i];
        }
        while(true){
            system("cls");
            switch(menu()){
                case 1:
                    system("cls");
                    cout<<"--Check if the array is sorted in descending order.--"<<endl;
                    cout<<"Your array is [ ";
                    for(int i=0; i<len; i++){
                        cout<<arr[i]<<" ";
                        if (i<len-1)
                            cout<<", ";
                    } cout<<"]\n";
                    if(sortcheck.iterative_sorted(arr, len))
                        cout<<"The array is sorted in descending order."<<endl;
                    else
                        cout<<"The array is not sorted in descending order."<<endl;
                    system("pause");
                    break;

                case 2:
                    system("cls");
                    cout<<"--recursive method to count the occurrence--"<<endl;
                    cout<<"Your array is [ ";
                    for(int i=0; i<len; i++){
                        cout<<arr[i]<<" ";
                        if (i<len-1)
                            cout<<", ";
                    } cout<<"]\n";
                    cout<<"Enter target number: ";
                    cin>>target;
                    cout<<"The number "<<target<<" Occurred "<<occurance.r_occurance_number(arr, len, target)<<" times!"<<endl;
                    system("pause");
                    break;

                case 3:
                    system("cls");
                    cout<<"--iterative method to find the smallest--"<<endl;
                    cout<<"Your array is [ ";
                    for(int i=0; i<len; i++){
                        cout<<arr[i]<<" ";
                        if (i<len-1)
                            cout<<", ";
                    } cout<<"]\n";
                    cout<<"The smallest element in your array is "<<small.i_smallest(arr, len)<<endl;
                    system("pause");
                    break;

                case 4:
                    system("cls");
                    cout<<"--Recursive method to find the smallest--"<<endl;
                    cout<<"Your array is [ ";
                    for(int i=0; i<len; i++){
                        cout<<arr[i]<<" ";
                        if (i<len-1)
                            cout<<", ";
                    } cout<<"]\n";
                    cout<<"The smallest element in your array is "<<small.r_smallest(arr, len)<<endl;
                    system("pause");
                    break;

                case 5:
                    system("cls");
                    cout<<"--Check if Palindrome Iterative Method--"<<endl;
                    cout<<"Your array is [ ";
                    for(int i=0; i<len; i++){
                        cout<<arr[i]<<" ";
                        if (i<len-1)
                            cout<<", ";
                    } cout<<"]\n";
                    if(palindromechecker.i_palindrome(arr, len))
                        cout<<"The array is a palindrome!"<<endl;
                    else
                        cout<<"The array is not a palindrome!"<<endl;
                    system("pause");
                    break;

                case 6:{
                    system("cls");
                    cout<<"--Check if Palindrome Recursive Method--"<<endl;
                    cout<<"Your array is [ ";
                    for(int i=0; i<len; i++){
                        cout<<arr[i]<<" ";
                        if (i<len-1)
                            cout<<", ";
                    } cout<<"]\n";
                    if(palindromechecker.r_palindrome(arr, 0, len-1))
                        cout<<"The array is a palindrome!"<<endl;
                    else
                        cout<<"The array is not a palindrome!"<<endl;
                    system("pause");
                    break;
                }

                case 7:
                    system("cls");
                    cout<<"--Enter New Array--"<<endl;
                    cout<<"Enter size of your array (max 10): ";
                    cin>>len;
                    cout<<"Enter elements: ";
                    for(int i=0; i<len; i++) {
                        cin>>arr[i];
                    }
                    break;

                case 8:
                    exit(0);
                    break;

                default:
                    cout<<"Invalid Input!"<<endl;
                    system("pause");
            }
        }
    }
};

int main(){
    Practice_Problems cs;
    cs.cases();

    return 0;
}

File: Python/Projects/practice_problems.py
class ISorted:
    def iterative_sorted(self, arr):
        return all(arr[i] >= arr[i + 1] for i in range(len(arr) - 1))


class ROccurrence:
    def r_occurance_number(self, arr, target):
        if not arr:
            return 0
        return (1 if arr[-1] == target else 0) + self.r_occurance_number(arr[:-1], target)


class Smallest:
    def i_smallest(self, arr):
        if not arr:
            return None
        return min(arr)

    def r_smallest(self, arr):
        if not arr:
            return None
        if len(arr) == 1:
            return arr[0]
        min_val = self.r_smallest(arr[:-1])
        return min(min_val, arr[-1])


class Palindrome:
    def i_palindrome(self, arr):
        return arr == arr[::-1]

    def r_palindrome(self, arr, left, right):
        if left >= right:
            return True
        if arr[left] != arr[right]:
            return False
        return self.r_palindrome(arr, left + 1, right - 1)


class PracticeProblems:
    def __init__(self):
        self.sortcheck = ISorted()
        self.occurance = ROccurrence()
        self.small = Smallest()
        self.palindromechecker = Palindrome()

    def menu(self):
        print("----------------------------Practice Problems-----------------------------")
        print("1. Iterative method to determine if array is sorted in descending order.")
        print("2. Recursive method to count the occurrence of a number in an array")
        print("3. Iterative method to find the smallest element in an array")
        print("4. Recursive method to find the smallest element in an array")
        print("5. Iterative method to check if the array is PALINDROME")
        print("6. Recursive method to check if the array is PALINDROME")
        print("7. Change the array")
        print("8. Exit")
        print("--------------------------------------------------------------------------")
        opt = int(input("[1-8]-> "))
        return opt

    def input_array(self, length):
        arr = []
        for i in range(length):
            val = int(input(f"Element {i + 1}: "))
            arr.append(val)
        return arr

    def cases(self):
        arr = []
        len_arr = 0

        while True:
            if len_arr == 0:
                len_arr = int(input("Enter size of your array (max 10): "))
                arr = self.input_array(len_arr)

            option = self.menu()
            if option == 1:
                print("--Check if the array is sorted in descending order.--")
                print(f"Your array is {arr}")
                if self.sortcheck.iterative_sorted(arr):
                    print("The array is sorted in descending order.")
                else:
                    print("The array is not sorted in descending order.")
            elif option == 2:
                print(f"Your array is {arr}")
                target = int(input("Enter target number: "))
                print(f"The number {target} occurred {self.occurance.r_occurance_number(arr, target)} times!")
            elif option == 3:
                print(f"Your array is {arr}")
                print(f"The smallest element in your array is {self.small.i_smallest(arr)}")
            elif option == 4:
                print(f"Your array is {arr}")
                print(f"The smallest element in your array is {self.small.r_smallest(arr)}")
            elif option == 5:
                print(f"Your array is {arr}")
                if self.palindromechecker.i_palindrome(arr):
                    print("The array is a palindrome!")
                else:
                    print("The array is not a palindrome!")
            elif option == 6:
                print(f"Your array is {arr}")
                if self.palindromechecker.r_palindrome(arr, 0, len(arr) - 1):
                    print("The array is a palindrome!")
                else:
                    print("The array is not a palindrome!")
            elif option == 7:
                len_arr = int(input("Enter size of your array (max 10): "))
                arr = self.input_array(len_arr)
            elif option == 8:
                break
            else:
                print("Invalid Input!")

if __name__ == "__main__":
    practice_problems = PracticeProblems()
    practice_problems.cases()
